cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java

```diff
@@ -169,14 +162,20 @@ public ApplicationId start(NamespaceId namespace, AppRequest<?> appRequest) thro
     Set<String> realDatasets = appRequest.getPreview() == null ? new HashSet<String>()
       : appRequest.getPreview().getRealDatasets();
 
-    ApplicationId previewApp = namespace.app(RunIds.generate().getId());
+    ApplicationId previewApp = namespace.app(PREFIX + System.currentTimeMillis());
     Injector injector = createPreviewInjector(previewApp, realDatasets);
-    appInjectors.put(previewApp, injector);
     PreviewRunner runner = injector.getInstance(PreviewRunner.class);
     if (runner instanceof Service) {
       ((Service) runner).startAndWait();
     }
-    runner.startPreview(new PreviewRequest<>(getProgramIdFromRequest(previewApp, appRequest), appRequest));
+    try {
+      runner.startPreview(new PreviewRequest<>(getProgramIdFromRequest(previewApp, appRequest), appRequest));
+    } catch (Exception e) {
+      LOG.error("Failed to start the preview run");
+      removePreviewDir(previewApp);
```

Comment:
I think we should also do `PreviewRunner.stop()` here to stop the services started by preview run.